docs/contributing/javascript_guidelines.rst

```diff
@@ -1,51 +1,179 @@
+    });
 
-Under the hood, the tasks use the `JavaScript Code Style <http://jscs.info/>`_ library.
+Note that we always wrap single arguments with brackets.
 
-To edit the settings for ignored files, or to change the linting rules,
-edit the ``.jscsrc`` file in the wagtail project root.
+**Spread operators (instead of Object.assign)**
 
-A complete list of the possible linting rules can be found here:
-`JSCS Rules <http://jscs.info/rules.html>`_
+.. code-block:: javascript
+
+    // Bad
+    case 'LOGIN_SUCCESS':
+      return Object.assign({}, state, {
+        email: action.payload.email,
+        profile: action.payload.profile
+      });
+
+    // Good
+    case 'LOGIN_SUCCESS':
+      return { ...state, email: action.payload.email, profile: action.payload.profile };
+
+**JSX**
+
+All components in Wagtail are written with the `JSX <https://facebook.github.io/react/docs/jsx-in-depth.html>`_ syntax.
+Aligment styles are applied as you will see below.
+
+.. code-block:: javascript
+
+    // Bad
+    <Component/>
+
+    // Bad
+    <Component myProperty="foo"
+               otherProperty="bar" />
+
+    // Good
+    <Component />
+
+    // Good
+    <Component
+      myProperty="foo"
+      otherProperty="bar"
+    />
+
+    // Good (if a single property fits on one line)
+    <Component myProperty="foo" />
+
+
+**Directory structures**
+
+In Wagtail all components have their own directory (placed in ``client/src/components``), documentation and styles.
+A component directory exists at least of the following files (e.g. explorer):
+
+* ``Explorer.js`` (since the component it's named ``Explorer``)
+* ``style.scss``
+* ``README.md``
+
+If the component is reusable we give them an own directory. It the component only belongs to one other component,
+place them in the same directory, e.g. ``ExplorerButton.js`` and ``Explorer.js``.
+
+**Actions**
+
+All actions within Wagtail are created by the `createAction` method of `redux-actions <https://github.com/acdlite/redux-actions>`_.
+
+Actions must be formatted as a `Flux Standard Action <https://github.com/acdlite/flux-standard-action>`_. In this case
+an action exists of the following properties.
+
+.. code-block:: javascript
+
+    {
+      payload: {
+        email: 'foo@bar.com'
+      },
+      type: 'LOGIN_SUCCESS'
+    }
+
+In case of an error, the action should be as you will see below (note that the ``payload`` should be an error object).
+
+.. code-block:: javascript
+
+    {
+      error: true,
+      payload: new Error('Sorry, the credentials you entered are invalid.'),
+      type: 'LOGIN_FAILURE'
+    }
+
+An action must not include properties other than ``type``, ``payload``, ``error``, and ``meta``. Since we are using
+`redux-action`, these requirements on actions are taken care for us.
+
+Also notice the ``_SUCCESS`` and ``_FAILURE`` appendix in the action name, names with ``_COMPLETE`` or ``_ERROR``
+**must not** be used instead.
+
+**Presentational and container components**
```

Comment:
If we intend to keep the `npm run component` command, we should double down on it (maybe use a yeoman generator?) and use it to generate those components with the "right" code style / architecture.

Here would be an appropriate place to mention this command.